src/reader/mod.rs

```diff
@@ -1930,3 +1933,68 @@ mod test {
     pub(super) use check;
     pub(super) use small_buffers;
 }
+
+fn read_to_string_impl<R: BufRead>(starting_buffer_position: usize, reader: &mut R, parser: Parser, end: QName, buf: &mut Vec<u8>) -> Result<Vec<u8>> {
+    // Initialize a new XMLParser
+    let reader = StoredReader::new(
+        reader, starting_buffer_position
+    );
+    let mut temp_reader: Reader<StoredReader<&mut R>> = Reader {
+        reader,
+        parser
+    };
+
+    let range = temp_reader.read_to_end_into(end, buf)?;
+    Ok(temp_reader.into_inner().get_stored_value(range.start, range.end))
+}
+
+/// Wraps an existing Reader and stores all the bytes read out of the `inner`.
+struct StoredReader<R> {
```

Comment:
This struct seems a reimplementation of `BufReader` with an additional `initial_offset` field. Is it possible to replace it by
```rust
struct StoredReader<R> {
    inner: BufReader<R>,
    initial_offset: usize,
}
```
?

Also, I would like to see some internal tests of just this reader that it work as expected. For example, that it will not consume more than required.

Reply:
Yes that's right, but we want to use a user-provided storage buffer. We can't do this with BufReader as BufReader will allocate it's own buffer.